### DISCUSSION TITLE
Freeform and 1:1 translations mixed - best practice?

### DISCUSSION THREAD

**Discussion by benjaoming** [wagtail/wagtail#7811] [General]
Hi all good Wagtail practitioners!

With an organizing that I help on a voluntary basis, we need to mix 1:1 translation with Freeform trees.

It's a little hard to explain without an example (graph below). Let's say our organization has a completely individual Contact page based on language -- it will explain the local office's contacts. But it also has a database of for instance products, and it's important that the product properties do not diverge.

For instance, we don't want "freeform" such that one translator can say that the product is blue and the other can specify red. We don't want this in the database at all, not even as a theoretical option. So we would like to use "django-modeltranslation" for this approach.

![image](https://user-images.githubusercontent.com/374612/147513310-37fa08d0-32c8-4ac5-be79-148b255b7b22.png)

The freeform-based Wagtail site would look like this:

![image](https://user-images.githubusercontent.com/374612/147513329-e45af412-ce44-40bd-82e0-663518e9ff0a.png)

The problem:

We started using `locale_id` and `translation_key` to have freeform trees, but now we run into issues: Using django-modeltranslation, we want our - say - product database as a model `ProductIndex(Page)` with `Product(Page)` nested in that tree. But with freeform translation trees, we would have to create a `ProductIndex` for each `locale_id` with its own set of Product objects. That doesn't work, because then all the products will end up in separately maintained nodes, not 1:1 translations.

If I could leave `locale_id` as "null", I would be able to create a modeltranslation-based `ProductIndex` with all the products created and translated inside Wagtail, but not inside a language's root ("Home (en)" etc). But lacking such options, we would have to manage and translate such a product database outside of Wagtail.

I see two possible solutions:

1) I cannot use Wagtail's `locale_id` but have to revert to the old way of doing things, with an explicit tree for each language and 1:1 pages nested outside those trees.

2) We don't use Wagtail for 1:1 translations but for instance django-admin or our own UI.

Has anyone else managed to have 1:1 translation inside Wagtail mixed with the current freeform-approach? Is there a clever hack to place such a ProductIndex outside of the freeform tree and still resolve according to a `/<language_slug>/` URL?

Thanks for reading,
Ben

p.s. I maintain [benjaoming/django-modeltranslation-wagtail](https://github.com/benjaoming/django-modeltranslation-wagtail) which has been a way of using django-modeltranslation in Wagtail for the 1:1 translation part. It's a simpler alternative to infoportugal/wagtail-modeltranslation. Previously, I built sites with a proxy root, "RootLanguageRedirect", it was the way to get freeform-translation working before `locale_id` and `translation_key`.

**Comment by allcaps** [comment]
Make `Product` regular Django model and decorate it as a snippet. The `ProductIndex` can have a `RoutablePageMixin` so that products can be served on a sub path.

This way you have one instance per product and can use model translation to define the translatable product fields.

You can manage snippets via the Snippets menu OR use Wagtail ModelAdmin. AFAIK model translation 'duplicates' fields for each language. You can use those field names in the Wagtail ModelAdmin definition.

**Comment by benjaoming** [comment]
So... quite some time ago, I asked this question and thanks @allcaps for getting back! Since then, I learned more about wagtail-localize by experience. I understood that my question about the previous 1:1 and freeform scenarios is no long well-posed. Simply put, these are not ultimately the only two types of scenarios. Because of the automatic sync'ing features of wagtail-localize, we will be in both scenarios at the same time and not really have to worry about which one nor bother the translators with this. Instead, we focus on workflows.

I tried to recap in the lightning talk starting around 9:20: https://www.youtube.com/watch?v=lAXjc9Jk5Ho
